File: Study Plan/LeetCode 75/1431 Kids With the Greatest Number of Candies/KidsWithTheGreatestNumberOfCandiesV2.java
/*  
    Problem
    --------------------------------------------------------------------------------
    1431. Kids With the Greatest Number of Candies
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 1 ms (Beats 88.74% of users with Java)
    - Memory 41.97 MB (Beats 83.93% of users with Java)
*/

class Solution {
    public List<Boolean> kidsWithCandies(int[] candies, int extraCandies) {
        List<Boolean> result = new ArrayList<>(candies.length);
        // Hint 1
        // Use greedy approach. For each kid check if 
        // candies[i] + extraCandies ≥ maximum in Candies[i].
        int maxCandies = Integer.MIN_VALUE;
        for (int candy : candies) {
            maxCandies = Math.max(maxCandies, candy);
        }
        for (int candy : candies) {
            result.add(candy + extraCandies >= maxCandies);
        }
        return result;
    }
}

File: Study Plan/LeetCode 75/1431 Kids With the Greatest Number of Candies/KidsWithTheGreatestNumberOfCandiesV2.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    1431. Kids With the Greatest Number of Candies
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/kids-with-the-greatest-number-of-candies/solutions/4384104/c-based-on-hints-beats-82-in-runtime-time-and-space-complexity-o-n/

    - Runtime 107 ms (Beats 82.50% of users with C#)
    - Memory 48.71 MB (Beats 9.32% of users with C#)
*/

public class Solution
{
    public IList<bool> KidsWithCandies(int[] candies, int extraCandies)
    {
        var result = new List<bool>(candies.Length);

        // Hint 1
        // Use greedy approach. For each kid check if 
        // candies[i] + extraCandies ≥ maximum in Candies[i].
        var maxCandies = candies.Max();

        foreach (var candy in candies)
        {
            result.Add(candy + extraCandies >= maxCandies);
        }

        return result;
    }
}